File: frontend/src/services/backtests.ts
export type BacktestKind = 'SIGNAL' | 'PORTFOLIO' | 'EXECUTION'
export type BacktestStatus = 'PENDING' | 'RUNNING' | 'COMPLETED' | 'FAILED'

export type UniverseSymbol = {
  symbol: string
  exchange?: string
}

export type BacktestUniverse = {
  mode: 'HOLDINGS' | 'GROUP' | 'BOTH'
  broker_name?: 'zerodha' | 'angelone'
  group_id?: number | null
  symbols?: UniverseSymbol[]
}

export type BacktestRun = {
  id: number
  owner_id?: number | null
  kind: BacktestKind
  status: BacktestStatus | string
  title?: string | null
  config: Record<string, unknown>
  result?: Record<string, unknown> | null
  error_message?: string | null
  started_at?: string | null
  finished_at?: string | null
  created_at: string
  updated_at: string
}

async function readApiError(res: Response): Promise<string> {
  const contentType = res.headers.get('content-type') ?? ''
  if (contentType.includes('application/json')) {
    try {
      const data = (await res.json()) as { detail?: unknown }
      if (typeof data.detail === 'string' && data.detail.trim()) return data.detail
      return JSON.stringify(data)
    } catch {
      // fall through
    }
  }

  try {
    const body = await res.text()
    return body.trim()
  } catch {
    return ''
  }
}

export async function createBacktestRun(payload: {
  kind: BacktestKind
  title?: string | null
  universe: BacktestUniverse
  config: Record<string, unknown>
}): Promise<BacktestRun> {
  const res = await fetch('/api/backtests/runs', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(payload),
  })
  if (!res.ok) {
    const detail = await readApiError(res)
    throw new Error(
      `Failed to create backtest run (${res.status})${detail ? `: ${detail}` : ''}`,
    )
  }
  return (await res.json()) as BacktestRun
}

export async function listBacktestRuns(params?: {
  kind?: BacktestKind
  limit?: number
}): Promise<BacktestRun[]> {
  const url = new URL('/api/backtests/runs', window.location.origin)
  if (params?.kind) url.searchParams.set('kind', params.kind)
  if (params?.limit != null) url.searchParams.set('limit', String(params.limit))
  const res = await fetch(url.toString(), { cache: 'no-store' })
  if (!res.ok) {
    const detail = await readApiError(res)
    throw new Error(
      `Failed to load backtest runs (${res.status})${detail ? `: ${detail}` : ''}`,
    )
  }
  return (await res.json()) as BacktestRun[]
}

export async function getBacktestRun(id: number): Promise<BacktestRun> {
  const res = await fetch(`/api/backtests/runs/${id}`, { cache: 'no-store' })
  if (!res.ok) {
    const detail = await readApiError(res)
    throw new Error(
      `Failed to load backtest run (${res.status})${detail ? `: ${detail}` : ''}`,
    )
  }
  return (await res.json()) as BacktestRun
}

File: backend/app/schemas/backtests.py
from __future__ import annotations

import json
from datetime import datetime
from typing import Any, Literal, Optional

from pydantic import BaseModel, Field

BacktestKind = Literal["SIGNAL", "PORTFOLIO", "EXECUTION"]
BacktestStatus = Literal["PENDING", "RUNNING", "COMPLETED", "FAILED"]


class UniverseSymbol(BaseModel):
    symbol: str = Field(min_length=1)
    exchange: str = Field(default="NSE", min_length=1)


class BacktestUniverse(BaseModel):
    mode: Literal["HOLDINGS", "GROUP", "BOTH"] = "GROUP"
    broker_name: str = "zerodha"
    group_id: Optional[int] = None
    symbols: list[UniverseSymbol] = Field(default_factory=list)


class BacktestRunCreate(BaseModel):
    kind: BacktestKind
    title: Optional[str] = None
    universe: BacktestUniverse = Field(default_factory=BacktestUniverse)
    config: dict[str, Any] = Field(default_factory=dict)


class BacktestRunRead(BaseModel):
    id: int
    owner_id: Optional[int] = None
    kind: str
    status: str
    title: Optional[str] = None
    config: dict[str, Any]
    result: Optional[dict[str, Any]] = None
    error_message: Optional[str] = None
    started_at: Optional[datetime] = None
    finished_at: Optional[datetime] = None
    created_at: datetime
    updated_at: datetime

    @classmethod
    def from_model(cls, obj) -> "BacktestRunRead":
        config = {}
        result = None
        try:
            config = json.loads(obj.config_json or "{}")
        except Exception:
            config = {}
        if obj.result_json:
            try:
                result = json.loads(obj.result_json)
            except Exception:
                result = None

        return cls(
            id=obj.id,
            owner_id=obj.owner_id,
            kind=obj.kind,
            status=obj.status,
            title=obj.title,
            config=config,
            result=result,
            error_message=obj.error_message,
            started_at=obj.started_at,
            finished_at=obj.finished_at,
            created_at=obj.created_at,
            updated_at=obj.updated_at,
        )


class EodCandleLoadRequest(BaseModel):
    symbols: list[UniverseSymbol] = Field(min_items=1)
    start: datetime
    end: datetime
    allow_fetch: bool = True


class EodCandleLoadResponse(BaseModel):
    dates: list[str]
    prices: dict[str, list[Optional[float]]]
    missing_symbols: list[str] = Field(default_factory=list)


class BacktestRunsDeleteRequest(BaseModel):
    ids: list[int] = Field(min_items=1)


class BacktestRunsDeleteResponse(BaseModel):
    deleted_ids: list[int] = Field(default_factory=list)
    forbidden_ids: list[int] = Field(default_factory=list)
    missing_ids: list[int] = Field(default_factory=list)


__all__ = [
    "BacktestKind",
    "BacktestRunCreate",
    "BacktestRunRead",
    "BacktestRunsDeleteRequest",
    "BacktestRunsDeleteResponse",
    "BacktestStatus",
    "BacktestUniverse",
    "EodCandleLoadRequest",
    "EodCandleLoadResponse",
    "UniverseSymbol",
]
